transaction/extra/src/memo/defragmentation.rs

```diff
@@ -0,0 +1,144 @@
+    /// If the fee exceeds 56 bits, an error will be thrown
+    pub fn new(
+        fee: u64,
+        total_outlay: u64,
+        defrag_id: u64,
+    ) -> Result<Self, DefragmentationMemoError> {
+        let mut result = Self {
+            fee: 0,
+            total_outlay,
+            defrag_id,
+        };
+        result.set_fee(fee)?;
+        Ok(result)
+    }
+
+    /// Returns the fee recorded in this memo
+    pub fn get_fee(&self) -> u64 {
+        self.fee
+    }
+
+    /// Sets the fee recorded in this memo
+    /// If the value given cannot be represented in 56 bits, an error is thrown
+    pub fn set_fee(&mut self, value: u64) -> Result<(), DefragmentationMemoError> {
+        if value.to_be_bytes()[0] != 0u8 {
+            return Err(DefragmentationMemoError::FeeTooLarge);
+        }
+        self.fee = value;
+        Ok(())
+    }
+
+    /// Returns the total outlay of the defragmentation transaction
+    pub fn get_total_outlay(&self) -> u64 {
+        self.total_outlay
+    }
+
+    /// Sets the total outlay
+    pub fn set_total_outlay(&mut self, value: u64) {
+        self.total_outlay = value;
+    }
+
+    /// Returns the defragmentation ID
+    pub fn get_defrag_id(&self) -> u64 {
+        self.defrag_id
+    }
+
+    /// Sets the defragmentation ID
+    pub fn set_defrag_id(&mut self, value: u64) {
+        self.defrag_id = value;
+    }
+}
+
+impl From<&[u8; DefragmentationMemo::MEMO_DATA_LEN]> for DefragmentationMemo {
+    // The layout of the memo data in 64 bytes is:
+    // [0-7): fee
+    // [7-15): total outlay
+    // [15-23): defrag ID
+    // [23-64) unused
+    fn from(src: &[u8; DefragmentationMemo::MEMO_DATA_LEN]) -> Self {
+        let fee = {
+            let mut fee_bytes = [0u8; 8];
+            fee_bytes[1..].copy_from_slice(&src[..7]);
+            u64::from_be_bytes(fee_bytes)
+        };
+        let total_outlay =
+            u64::from_be_bytes(src[7..15].try_into().expect("BUG! arithmetic error"));
+        let defrag_id = u64::from_be_bytes(src[15..23].try_into().expect("BUG! arithmetic error"));
+        Self {
+            fee,
+            total_outlay,
+            defrag_id,
+        }
+    }
+}
+
+impl From<DefragmentationMemo> for [u8; DefragmentationMemo::MEMO_DATA_LEN] {
+    fn from(src: DefragmentationMemo) -> [u8; DefragmentationMemo::MEMO_DATA_LEN] {
+        let mut memo_data = [0u8; DefragmentationMemo::MEMO_DATA_LEN];
+        memo_data[..7].copy_from_slice(&src.fee.to_be_bytes()[1..]);
+        memo_data[7..15].copy_from_slice(&src.total_outlay.to_be_bytes());
+        memo_data[15..23].copy_from_slice(&src.defrag_id.to_be_bytes());
+        memo_data
+    }
+}
+
+/// An error that can occur when configuring a destination memo
+#[derive(Display, Debug)]
+pub enum DefragmentationMemoError {
+    /// The fee amount is too large to be represented in the DefragmentationMemo
+    FeeTooLarge,
+}
+
+impl_memo_type_conversions! { DefragmentationMemo }
```

Comment:
:wrench: Not sure why the destination memo test is in `memo/mod.rs` but suggest putting the test for the defragmentation memo in here.

There are a few changes that may not be agreeable so take anything you like:
- remove name prefix of `test_` that's generally used for discovery for xunit frameworks while the `#[test]` macro in rust provides the discovery
- Broke the tests out to separate test functions. We have quite a few large tests so would like to see smaller more focused ones. The test stops on the first failure so if multiple things would fail you only get the first one.
- Explicit comparison against the expected error, this requires adding `PartialEq` to the Error derives
- Dropped `u64` from the literals, fetl kind of noisy, but explicit is nice :shrug: 
- Used a constant for the boundary tests to max fee
- Removed the bitwise not on the id usage, it's felt like it might be easy to miss.
```suggestion
impl_memo_type_conversions! { DefragmentationMemo }

#[cfg(test)]
mod test {
    use super::*;

    const MAX_FEE: u64 = 0x00FFFFFF_FFFFFFFF;

    #[test]
    fn getters_and_setters() {
        let mut memo = DefragmentationMemo::new(48237, 9001, 3405697037).unwrap();

        assert_eq!(48237, memo.get_fee());
        assert_eq!(9001, memo.get_total_outlay());
        assert_eq!(3405697037, memo.get_defrag_id());

        memo.set_fee(73284).unwrap();
        memo.set_total_outlay(8999);
        memo.set_defrag_id(10);

        assert_eq!(73284, memo.get_fee());
        assert_eq!(8999, memo.get_total_outlay());
        assert_eq!(10, memo.get_defrag_id());
    }

    #[test]
    fn max_fee_succeeds() {
        let memo = DefragmentationMemo::new(MAX_FEE, 0, 0).unwrap();
        assert_eq!(MAX_FEE, memo.get_fee());
    }

    #[test]
    fn exceeding_max_fee_fails() {
        let mut memo = DefragmentationMemo::new(0, 0u64, 0u64).unwrap();
        assert_eq!(
            memo.set_fee(MAX_FEE + 1),
            Err(DefragmentationMemoError::FeeTooLarge)
        );
    }

    #[test]
    fn new_propagates_error() {
        assert_eq!(
            DefragmentationMemo::new(MAX_FEE + 1, 0u64, 0u64),
            Err(DefragmentationMemoError::FeeTooLarge)
        );
    }

    #[test]
    fn serialization() {
        let memo = DefragmentationMemo::new(2525, 36, 80).unwrap();

        let memo_bytes: [u8; DefragmentationMemo::MEMO_DATA_LEN] = memo.clone().into();
        let deserialized_memo = DefragmentationMemo::from(&memo_bytes);

        assert_eq!(memo, deserialized_memo);
    }
}

```